File: agenkit-ts/src/protocols/agui/hitl.ts
/**
 * AG-UI Human-in-the-Loop Integration
 *
 * Integrates the HumanInLoopAgent pattern with AG-UI protocol using Interrupt events.
 * Provides streaming approval workflow where agents can request human approval via
 * Interrupt events, and frontends can respond with InterruptResponse messages.
 *
 * Key concepts:
 * - Interrupt events for approval requests
 * - InterruptResponse for approval decisions
 * - Streaming approval workflow
 * - Integration with HumanInLoopAgent pattern
 *
 * Example:
 *   import { AGUIHumanInLoopAdapter } from './agui/hitl';
 *   import { HumanInLoopAgent } from '../patterns/human-in-loop';
 *
 *   // Create human-in-loop agent
 *   const hilAgent = new HumanInLoopAgent({
 *     agent: myAgent,
 *     approvalFunc: myApprovalFunc,
 *     approvalThreshold: 0.8
 *   });
 *
 *   // Wrap with AG-UI adapter
 *   const adapter = new AGUIHumanInLoopAdapter(hilAgent);
 *
 *   // Stream events (includes Interrupt events for approval requests)
 *   for await (const event of adapter.streamEvents(userMessage)) {
 *     if (event instanceof Interrupt) {
 *       // Frontend displays approval request
 *       // User responds via InterruptResponse
 *     }
 *   }
 */

import { Agent, Message } from '../../core/interfaces.js';
import { HumanInLoopAgent, ApprovalRequest } from '../../patterns/human-in-loop.js';
import { AGUIAdapter, AGUIAdapterConfig } from './adapter.js';
import {
  AGUIEvent,
  Interrupt,
  InterruptAction,
  InterruptReason,
  InterruptResponse,
  MetadataEvent,
  TextMessageChunk,
  TextMessageComplete,
  TextMessageStart,
} from './events.js';
import { randomUUID } from 'crypto';

/**
 * Configuration for AG-UI HITL adapter
 */
export interface AGUIHumanInLoopConfig extends AGUIAdapterConfig {
  /** Whether to emit Interrupt events for approval requests */
  emitInterrupts?: boolean;
}

/**
 * AG-UI adapter with human-in-the-loop support via Interrupt events.
 *
 * This adapter integrates the HumanInLoopAgent pattern with AG-UI protocol.
 * When an agent requires approval (confidence < threshold), an Interrupt event
 * is emitted to request human approval. The frontend can respond via
 * InterruptResponse.
 *
 * The adapter handles:
 * - Converting approval requests to Interrupt events
 * - Processing InterruptResponse from frontend
 * - Streaming approval workflow
 * - Metadata about approval decisions
 *
 * Example:
 *   const hilAgent = new HumanInLoopAgent({
 *     agent: myAgent,
 *     approvalFunc: async (request) => {
 *       // Custom approval logic
 *       return { approved: true };
 *     },
 *     approvalThreshold: 0.8
 *   });
 *
 *   const adapter = new AGUIHumanInLoopAdapter(hilAgent);
 *
 *   for await (const event of adapter.streamEvents(message)) {
 *     if (event instanceof Interrupt) {
 *       // Display approval request to user
 *       console.log(`Approval needed: ${event.message}`);
 *       console.log(`Confidence: ${event.context.confidence}`);
 *     }
 *   }
 */
export class AGUIHumanInLoopAdapter extends AGUIAdapter {
  private readonly emitInterrupts: boolean;
  private readonly pendingInterrupts: Map<string, any>;
  private approvalRequestCallback?: (request: ApprovalRequest) => Promise<void>;

  /**
   * Initialize AG-UI human-in-loop adapter.
   *
   * @param agent - Agent to wrap (HumanInLoopAgent or regular Agent)
   * @param config - Optional configuration
   */
  constructor(agent: Agent, config: AGUIHumanInLoopConfig = {}) {
    super(agent, config);
    this.emitInterrupts = config.emitInterrupts !== false;
    this.pendingInterrupts = new Map();
  }

  /**
   * Stream AG-UI events with interrupt support.
   *
   * When the agent requires approval, emits an Interrupt event to notify
   * the frontend about the approval decision.
   *
   * Note: This implementation emits Interrupt events after the approval
   * decision has been made (informational). For true bidirectional HITL,
   * use a custom approvalFunc that integrates with your transport layer.
   *
   * @param message - Input message to process
   * @param messageId - Optional message ID
   * @param emitMetadata - Whether to emit MetadataEvent first
   * @yields AG-UI events (includes Interrupt events for approval notifications)
   *
   * Example:
   *   for await (const event of adapter.streamEvents(message)) {
   *     if (event instanceof Interrupt) {
   *       // Approval decision was made
   *       console.log(`Approval: ${event.context.approval_status}`);
   *     }
   *   }
   */
  async *streamEvents(
    message: Message,
    messageId?: string,
    emitMetadata: boolean = true,
  ): AsyncGenerator<AGUIEvent, void, undefined> {
    const msgId = messageId || this.generateMessageId();

    // Check if agent is a HumanInLoopAgent
    const agent = this.getAgent();
    const isHilAgent = agent instanceof HumanInLoopAgent;

    // For regular agents or if interrupts disabled, use standard streaming
    if (!isHilAgent || !this.emitInterrupts) {
      yield* super.streamEvents(message, msgId, emitMetadata);
      return;
    }

    // Emit metadata about agent capabilities (including HITL)
    if (emitMetadata) {
      yield this.createHITLMetadataEvent();
    }

    // Emit text message start
    yield new TextMessageStart('assistant', msgId, {
      agent_name: this.getAgentName(),
    });

    try {
      // Track if approval was requested
      let approvalRequested = false;
      let approvalDetails: any = {};

      // Store the original approval function
      const hilAgent = agent as HumanInLoopAgent;
      const originalApprovalFunc = (hilAgent as any).approvalFunc;

      // Wrap approval function to capture approval events
      (hilAgent as any).approvalFunc = async (request: ApprovalRequest) => {
        approvalRequested = true;
        approvalDetails = {
          confidence: request.confidence,
          timestamp: request.timestamp,
          context: request.context,
        };

        // Call original approval function
        return await originalApprovalFunc(request);
      };

      // Process message with agent
      const response = await agent.process(message);

      // Restore original approval function
      (hilAgent as any).approvalFunc = originalApprovalFunc;

      // If approval was requested, emit Interrupt event
      if (approvalRequested && this.emitInterrupts) {
        const interruptId = `interrupt_${randomUUID()}`;
        const interrupt = this.createApprovalInterrupt(
          interruptId,
          approvalDetails,
          response,
        );
        yield interrupt;
      }

      // Extract content
      const content =
        typeof response.content === 'string'
          ? response.content
          : JSON.stringify(response.content);

      // Stream content in chunks
      const chunkSize = (this as any).chunkSize || 50;
      for (let i = 0; i < content.length; i += chunkSize) {
        const chunk = content.slice(i, i + chunkSize);
        yield new TextMessageChunk(chunk, msgId, {
          chunk_index: Math.floor(i / chunkSize),
        });
      }

      // Emit completion
      yield new TextMessageComplete(content, 'stop', msgId, {
        agent_name: this.getAgentName(),
        response_metadata: response.metadata,
        approval_requested: approvalRequested,
        approval_details: approvalRequested ? approvalDetails : undefined,
      });
    } catch (error) {
      // Convert exceptions to error events
      yield this.createErrorEvent(msgId, error as Error);

      // Also emit a completion with error
      yield new TextMessageComplete('', 'error', msgId, {
        error: error instanceof Error ? error.message : String(error),
      });
    }
  }

  /**
   * Create metadata event with HITL capabilities
   */
  private createHITLMetadataEvent(): MetadataEvent {
    const agent = this.getAgent();
    const isHilAgent = agent instanceof HumanInLoopAgent;

    const metadata: Record<string, any> = {
      agent_name: this.getAgentName(),
      protocol: 'ag-ui',
      protocol_version: '1.0',
      capabilities: {
        streaming: true,
        tool_calls: false,
        interrupts: isHilAgent && this.emitInterrupts,
        multimodal: false,
      },
    };

    // Add HITL-specific metadata
    if (isHilAgent) {
      const hilAgent = agent as any;
      metadata.hitl = {
        enabled: true,
        approval_threshold: hilAgent.approvalThreshold || 0.8,
        confidence_key: hilAgent.confidenceKey || 'confidence',
      };
    }

    // Add agent capabilities if available
    if (agent.capabilities) {
      metadata.agent_capabilities = agent.capabilities();
    }

    return new MetadataEvent(metadata);
  }

  /**
   * Create an Interrupt event for approval request
   */
  private createApprovalInterrupt(
    interruptId: string,
    approvalDetails: any,
    response: Message,
  ): Interrupt {
    const confidence = approvalDetails.confidence || 0.0;
    const threshold = (this.getAgent() as any).approvalThreshold || 0.8;

    // Determine approval status
    const approved = confidence >= threshold;
    const status = approved ? 'approved' : 'rejected';

    // Create interrupt message
    const message = approved
      ? `Agent action approved (confidence: ${confidence.toFixed(2)})`
      : `Agent action requires approval (confidence: ${confidence.toFixed(2)}, threshold: ${threshold.toFixed(2)})`;

    return new Interrupt(
      InterruptReason.APPROVAL_REQUIRED,
      message,
      [InterruptAction.APPROVE, InterruptAction.REJECT, InterruptAction.EDIT],
      {
        approval_status: status,
        confidence: confidence,
        threshold: threshold,
        confidence_shortfall: Math.max(0, threshold - confidence),
        response_preview: String(response.content).slice(0, 100),
        timestamp: approvalDetails.timestamp,
        ...approvalDetails.context,
      },
      interruptId,
    );
  }

  /**
   * Generate a unique message ID
   */
  private generateMessageId(): string {
    return `msg_${randomUUID()}`;
  }

  /**
   * Create an error event from an exception
   */
  private createErrorEvent(messageId: string, error: Error): any {
    const ErrorEventClass = require('./events.js').ErrorEvent;
    return new ErrorEventClass(
      'agent_error',
      error.message || 'Unknown error',
      true,
      {
        message_id: messageId,
        error_type: error.name,
        stack: error.stack,
      },
    );
  }
}

/**
 * Convenience function to wrap a HumanInLoopAgent with AG-UI HITL adapter
 *
 * @param agent - HumanInLoopAgent to wrap
 * @param config - Optional configuration
 * @returns Wrapped AG-UI HITL adapter
 */
export function wrapHITLAgentAsAGUI(
  agent: HumanInLoopAgent,
  config?: AGUIHumanInLoopConfig,
): AGUIHumanInLoopAdapter {
  return new AGUIHumanInLoopAdapter(agent, config);
}

File: agenkit-rust/src/protocols/agui/hitl.rs
//! AG-UI Human-in-the-Loop Integration
//!
//! Integrates the HumanInLoopAgent pattern with AG-UI protocol using Interrupt events.
//! Provides streaming approval workflow where agents can request human approval via
//! Interrupt events, and frontends can respond with InterruptResponse messages.
use crate::core::{Agent, AgentError, Message};
use crate::patterns::human_in_loop::HumanInLoopAgent;
use crate::protocols::agui::adapter::{AGUIAdapter, AGUIAdapterConfig};
use crate::protocols::agui::events::*;
use futures::stream::{Stream, StreamExt};
use std::collections::HashMap;
use std::pin::Pin;
use std::sync::Arc;
use tokio::sync::mpsc;
use uuid::Uuid;

/// Configuration for AG-UI HITL adapter.
#[derive(Debug, Clone)]
pub struct AGUIHumanInLoopConfig {
    /// Base adapter configuration
    pub base_config: AGUIAdapterConfig,
    /// Whether to emit Interrupt events for approval requests
    pub emit_interrupts: bool,
}

impl Default for AGUIHumanInLoopConfig {
    fn default() -> Self {
        Self {
            base_config: AGUIAdapterConfig::default(),
            emit_interrupts: true,
        }
    }
}

/// AG-UI adapter with human-in-the-loop support via Interrupt events.
///
/// This adapter integrates the HumanInLoopAgent pattern with AG-UI protocol.
/// When an agent requires approval (confidence < threshold), an Interrupt event
/// is emitted to request human approval. The frontend can respond via
/// InterruptResponse.
///
/// The adapter handles:
/// - Converting approval requests to Interrupt events
/// - Processing InterruptResponse from frontend
/// - Streaming approval workflow
/// - Metadata about approval decisions
///
/// # Example
/// ```no_run
/// use agenkit::core::{Agent, Message};
/// use agenkit::patterns::{HumanInLoopAgent, HumanInLoopConfig, simple_approval_func};
/// use agenkit::protocols::agui::{AGUIHumanInLoopAdapter, AGUIHumanInLoopConfig, AGUIEvent, EventType};
/// use futures::StreamExt;
/// use std::sync::Arc;
///
/// # async fn example() -> Result<(), Box<dyn std::error::Error>> {
/// # let agent: Arc<dyn Agent> = todo!();
/// let hil_agent = HumanInLoopAgent::new(HumanInLoopConfig {
///     agent,
///     approval_threshold: 0.8,
///     approval_func: simple_approval_func(true),
///     confidence_key: "confidence".to_string(),
/// })?;
///
/// let adapter = AGUIHumanInLoopAdapter::new(
///     Arc::new(hil_agent),
///     AGUIHumanInLoopConfig::default()
/// );
///
/// let message = Message::with_text("user", "Process this action");
/// let mut stream = adapter.stream_events(message, None, true).await;
///
/// while let Some(event) = stream.next().await {
///     if event.event_type() == EventType::Interrupt {
///         // Display approval request to user
///         println!("Approval needed!");
///     }
/// }
/// # Ok(())
/// # }
/// ```
pub struct AGUIHumanInLoopAdapter {
    agent: Arc<dyn Agent>,
    base_adapter: AGUIAdapter,
    config: AGUIHumanInLoopConfig,
}

impl AGUIHumanInLoopAdapter {
    /// Create a new AG-UI human-in-loop adapter.
    ///
    /// # Arguments
    /// * `agent` - Agent to wrap (HumanInLoopAgent or regular Agent)
    /// * `config` - Adapter configuration
    pub fn new(agent: Arc<dyn Agent>, config: AGUIHumanInLoopConfig) -> Self {
        let base_adapter = AGUIAdapter::new(agent.clone(), config.base_config.clone());

        Self {
            agent,
            base_adapter,
            config,
        }
    }

    /// Stream AG-UI events with interrupt support.
    ///
    /// When the agent requires approval, emits an Interrupt event to notify
    /// the frontend about the approval decision.
    ///
    /// Note: This implementation emits Interrupt events after the approval
    /// decision has been made (informational). For true bidirectional HITL,
    /// use a custom approvalFunc that integrates with your transport layer.
    ///
    /// # Arguments
    /// * `message` - Input message to process
    /// * `message_id` - Optional message ID
    /// * `emit_metadata` - Whether to emit metadata event first
    ///
    /// # Returns
    /// Stream of AG-UI events (includes Interrupt events for approval notifications)
    pub async fn stream_events(
        &self,
        message: Message,
        message_id: Option<String>,
        emit_metadata: bool,
    ) -> Pin<Box<dyn Stream<Item = Box<dyn AGUIEvent>> + Send>> {
        let msg_id = message_id.unwrap_or_else(|| format!("msg_{}", Uuid::new_v4()));
        let agent = self.agent.clone();
        let agent_name = self
            .config
            .base_config
            .agent_name
            .clone()
            .unwrap_or_else(|| agent.name().to_string());
        let chunk_size = self.config.base_config.chunk_size;
        let emit_interrupts = self.config.emit_interrupts;

        // Check if agent is a HumanInLoopAgent
        let is_hil_agent = self.is_human_in_loop_agent();

        // For regular agents or if interrupts disabled, use standard streaming
        if !is_hil_agent || !emit_interrupts {
            return self
                .base_adapter
                .stream_events(message, Some(msg_id), emit_metadata)
                .await;
        }

        let (tx, mut rx) = mpsc::channel::<Box<dyn AGUIEvent>>(100);

        tokio::spawn(async move {
            // Emit metadata about agent capabilities (including HITL)
            if emit_metadata {
                let metadata_event = create_hitl_metadata_event(&agent_name, agent.as_ref());
                let _ = tx.send(Box::new(metadata_event)).await;
            }

            // Emit text message start
            let start_event = TextMessageStart::new("assistant", Some(msg_id.clone()))
                .with_metadata("agent_name", serde_json::json!(agent_name));
            let _ = tx.send(Box::new(start_event)).await;

            // Process message with agent
            match agent.process(message).await {
                Ok(response) => {
                    // Check if approval was requested by examining metadata
                    let approval_needed = response
                        .metadata
                        .get("approval_needed")
                        .and_then(|v| v.as_bool())
                        .unwrap_or(false);

                    let confidence = response
                        .metadata
                        .get("confidence")
                        .and_then(|v| v.as_f64())
                        .unwrap_or(0.0);

                    let approval_threshold = response
                        .metadata
                        .get("approval_threshold")
                        .and_then(|v| v.as_f64())
                        .unwrap_or(0.8);

                    let approval_status = response
                        .metadata
                        .get("approval_status")
                        .and_then(|v| v.as_str())
                        .unwrap_or("unknown");

                    // If approval was requested, emit Interrupt event
                    if approval_needed {
                        let interrupt_id = format!("interrupt_{}", Uuid::new_v4());
                        let interrupt = create_approval_interrupt(
                            interrupt_id,
                            confidence,
                            approval_threshold,
                            approval_status,
                            &response,
                        );
                        let _ = tx.send(Box::new(interrupt)).await;
                    }

                    // Extract content as string
                    let content = match response.content {
                        serde_json::Value::String(s) => s,
                        other => other.to_string(),
                    };

                    // Stream content in chunks
                    for (i, chunk) in content.as_bytes().chunks(chunk_size).enumerate() {
                        let chunk_str = String::from_utf8_lossy(chunk).to_string();
                        let chunk_event = TextMessageChunk::new(chunk_str, Some(msg_id.clone()))
                            .with_metadata("chunk_index", serde_json::json!(i));
                        let _ = tx.send(Box::new(chunk_event)).await;
                    }

                    // Emit complete event
                    let mut complete_event =
                        TextMessageComplete::new(content, "stop", Some(msg_id.clone()))
                            .with_metadata("agent_name", serde_json::json!(agent_name))
                            .with_metadata(
                                "approval_requested",
                                serde_json::json!(approval_needed),
                            );

                    // Add response metadata
                    if let Some(metadata) = serde_json::to_value(&response.metadata).ok() {
                        complete_event =
                            complete_event.with_metadata("response_metadata", metadata);
                    }

                    // Add approval details if present
                    if approval_needed {
                        complete_event = complete_event.with_metadata(
                            "approval_details",
                            serde_json::json!({
                                "confidence": confidence,
                                "threshold": approval_threshold,
                                "status": approval_status,
                            }),
                        );
                    }

                    let _ = tx.send(Box::new(complete_event)).await;
                }
                Err(error) => {
                    // Emit error event
                    let error_event = ErrorEvent::new(
                        "agent_error",
                        error.to_string(),
                        true,
                        Some(serde_json::json!({
                            "message_id": msg_id,
                            "error_type": format!("{:?}", error),
                        })),
                    );
                    let _ = tx.send(Box::new(error_event)).await;

                    // Emit complete with error finish reason
                    let complete_event =
                        TextMessageComplete::new("", "error", Some(msg_id.clone()))
                            .with_metadata("error", serde_json::json!(error.to_string()));
                    let _ = tx.send(Box::new(complete_event)).await;
                }
            }
        });

        Box::pin(async_stream::stream! {
            while let Some(event) = rx.recv().await {
                yield event;
            }
        })
    }

    /// Get the underlying agent.
    pub fn agent(&self) -> &Arc<dyn Agent> {
        &self.agent
    }

    /// Check if the agent is a HumanInLoopAgent by checking its capabilities.
    fn is_human_in_loop_agent(&self) -> bool {
        let caps = self.agent.capabilities();
        caps.contains(&"human-in-loop".to_string())
    }
}

/// Create metadata event with HITL capabilities.
fn create_hitl_metadata_event(agent_name: &str, agent: &dyn Agent) -> MetadataEvent {
    let mut data = HashMap::new();
    data.insert("agent_name".to_string(), serde_json::json!(agent_name));
    data.insert("protocol".to_string(), serde_json::json!("ag-ui"));
    data.insert("protocol_version".to_string(), serde_json::json!("1.0"));

    let caps = agent.capabilities();
    let is_hitl = caps.contains(&"human-in-loop".to_string());

    let mut capabilities = HashMap::new();
    capabilities.insert("streaming", serde_json::Value::Bool(true));
    capabilities.insert("tool_calls", serde_json::Value::Bool(false));
    capabilities.insert("interrupts", serde_json::Value::Bool(is_hitl));
    capabilities.insert("multimodal", serde_json::Value::Bool(false));

    data.insert(
        "capabilities".to_string(),
        serde_json::to_value(capabilities).unwrap_or(serde_json::Value::Null),
    );

    // Add HITL-specific metadata if available
    if is_hitl {
        // Note: In Rust, we can't easily extract private fields from HumanInLoopAgent
        // So we use default values or rely on the agent's introspection if available
        data.insert(
            "hitl".to_string(),
            serde_json::json!({
                "enabled": true,
                "approval_threshold": 0.8,
                "confidence_key": "confidence",
            }),
        );
    }

    // Add agent capabilities if available
    let introspection = agent.introspect();
    if !introspection.capabilities.is_empty() {
        data.insert(
            "agent_capabilities".to_string(),
            serde_json::to_value(&introspection.capabilities).unwrap_or(serde_json::Value::Null),
        );
    }

    MetadataEvent::new(data)
}

/// Create an Interrupt event for approval request.
fn create_approval_interrupt(
    interrupt_id: String,
    confidence: f64,
    threshold: f64,
    approval_status: &str,
    response: &Message,
) -> Interrupt {
    // Determine if approved based on status
    let approved = approval_status == "approved" || approval_status == "bypassed";

    // Create interrupt message
    let message = if approved {
        format!("Agent action approved (confidence: {:.2})", confidence)
    } else if approval_status == "rejected" {
        format!(
            "Agent action rejected (confidence: {:.2}, threshold: {:.2})",
            confidence, threshold
        )
    } else {
        format!(
            "Agent action requires approval (confidence: {:.2}, threshold: {:.2})",
            confidence, threshold
        )
    };

    // Extract content preview
    let content_preview = match &response.content {
        serde_json::Value::String(s) => {
            if s.len() > 100 {
                format!("{}...", &s[..100])
            } else {
                s.clone()
            }
        }
        other => {
            let s = other.to_string();
            if s.len() > 100 {
                format!("{}...", &s[..100])
            } else {
                s
            }
        }
    };

    let mut context = HashMap::new();
    context.insert(
        "approval_status".to_string(),
        serde_json::json!(approval_status),
    );
    context.insert("confidence".to_string(), serde_json::json!(confidence));
    context.insert("threshold".to_string(), serde_json::json!(threshold));
    context.insert(
        "confidence_shortfall".to_string(),
        serde_json::json!((threshold - confidence).max(0.0)),
    );
    context.insert(
        "response_preview".to_string(),
        serde_json::json!(content_preview),
    );
    context.insert(
        "timestamp".to_string(),
        serde_json::json!(response.timestamp.to_rfc3339()),
    );

    // Add any additional context from response metadata
    if let Some(agent_name) = response.metadata.get("agent") {
        context.insert("agent".to_string(), agent_name.clone());
    }

    Interrupt::new(
        InterruptReason::ApprovalRequired,
        message,
        vec![
            InterruptAction::Approve,
            InterruptAction::Reject,
            InterruptAction::Edit,
        ],
        context,
        Some(interrupt_id),
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::patterns::human_in_loop::{simple_approval_func, HumanInLoopConfig};
    use async_trait::async_trait;

    struct MockAgent {
        response: String,
        confidence: f64,
    }

    #[async_trait]
    impl Agent for MockAgent {
        fn name(&self) -> &str {
            "MockAgent"
        }

        async fn process(&self, _message: Message) -> Result<Message, AgentError> {
            Ok(
                Message::new("assistant", serde_json::json!(self.response.clone()))
                    .with_metadata("confidence", serde_json::json!(self.confidence)),
            )
        }
    }

    #[tokio::test]
    async fn test_hitl_adapter_streams_events() {
        let agent = Arc::new(MockAgent {
            response: "Test response".to_string(),
            confidence: 0.5,
        });

        let hil_agent = HumanInLoopAgent::new(HumanInLoopConfig {
            agent,
            approval_threshold: 0.8,
            approval_func: simple_approval_func(true),
            confidence_key: "confidence".to_string(),
        })
        .unwrap();

        let adapter =
            AGUIHumanInLoopAdapter::new(Arc::new(hil_agent), AGUIHumanInLoopConfig::default());

        let message = Message::new("user", serde_json::json!("test"));
        let mut stream = adapter.stream_events(message, None, false).await;

        let mut event_types = Vec::new();
        while let Some(event) = stream.next().await {
            event_types.push(event.event_type());
        }

        assert!(event_types.contains(&EventType::TextMessageStart));
        assert!(event_types.contains(&EventType::TextMessageChunk));
        assert!(event_types.contains(&EventType::TextMessageComplete));
        assert!(event_types.contains(&EventType::Interrupt)); // Should have interrupt for low confidence
    }

    #[tokio::test]
    async fn test_hitl_adapter_emits_metadata() {
        let agent = Arc::new(MockAgent {
            response: "Test".to_string(),
            confidence: 0.5,
        });

        let hil_agent = HumanInLoopAgent::new(HumanInLoopConfig {
            agent,
            approval_threshold: 0.8,
            approval_func: simple_approval_func(true),
            confidence_key: "confidence".to_string(),
        })
        .unwrap();

        let adapter =
            AGUIHumanInLoopAdapter::new(Arc::new(hil_agent), AGUIHumanInLoopConfig::default());

        let message = Message::new("user", serde_json::json!("test"));
        let mut stream = adapter.stream_events(message, None, true).await;

        let first_event = stream.next().await.unwrap();
        assert_eq!(first_event.event_type(), EventType::Metadata);

        // Check that metadata includes HITL capabilities
        let json = first_event.to_json();
        assert_eq!(json.get("protocol"), Some(&serde_json::json!("ag-ui")));
        assert!(json.get("hitl").is_some());
    }

    #[tokio::test]
    async fn test_hitl_adapter_high_confidence_bypasses() {
        let agent = Arc::new(MockAgent {
            response: "High confidence".to_string(),
            confidence: 0.95,
        });

        let hil_agent = HumanInLoopAgent::new(HumanInLoopConfig {
            agent,
            approval_threshold: 0.8,
            approval_func: simple_approval_func(true),
            confidence_key: "confidence".to_string(),
        })
        .unwrap();

        let adapter =
            AGUIHumanInLoopAdapter::new(Arc::new(hil_agent), AGUIHumanInLoopConfig::default());

        let message = Message::new("user", serde_json::json!("test"));
        let mut stream = adapter.stream_events(message, None, false).await;

        let mut has_interrupt = false;
        while let Some(event) = stream.next().await {
            if event.event_type() == EventType::Interrupt {
                has_interrupt = true;
            }
        }

        // High confidence should not trigger interrupt
        assert!(!has_interrupt);
    }

    #[tokio::test]
    async fn test_hitl_adapter_regular_agent_no_interrupts() {
        let agent = Arc::new(MockAgent {
            response: "Regular agent".to_string(),
            confidence: 0.5,
        });

        let adapter = AGUIHumanInLoopAdapter::new(agent, AGUIHumanInLoopConfig::default());

        let message = Message::new("user", serde_json::json!("test"));
        let mut stream = adapter.stream_events(message, None, false).await;

        let mut has_interrupt = false;
        while let Some(event) = stream.next().await {
            if event.event_type() == EventType::Interrupt {
                has_interrupt = true;
            }
        }

        // Regular agent (not HumanInLoopAgent) should not emit interrupts
        assert!(!has_interrupt);
    }
}
